ports/esp32/machine_pwm.c

```diff
@@ -119,74 +146,177 @@ STATIC void configure_channel(machine_pwm_obj_t *self) {
     }
 }
 
-STATIC void set_freq(int newval, ledc_timer_config_t *timer) {
-    // If already set, do nothing
-    if (newval == timer->freq_hz) {
-        return;
-    }
+STATIC void set_duty_u16(machine_pwm_obj_t *self, int duty);
+STATIC void set_duty_u10(machine_pwm_obj_t *self, int duty);
+STATIC void set_duty_ns(machine_pwm_obj_t *self, int ns);
 
-    // Find the highest bit resolution for the requested frequency
-    if (newval <= 0) {
-        newval = 1;
-    }
-    unsigned int res = 0;
-    for (unsigned int i = LEDC_APB_CLK_HZ / newval; i > 1; i >>= 1) {
-        ++res;
-    }
-    if (res == 0) {
-        res = 1;
-    } else if (res > PWRES) {
-        // Limit resolution to PWRES to match units of our duty
-        res = PWRES;
-    }
+STATIC void set_freq(unsigned int freq, ledc_timer_config_t *timer, machine_pwm_obj_t *self) {
```

Comment:
usually `self` is passed as the first arg

Reply:
Ok